### PR TITLE
Convert IBus.Subscribe and IBus.Unsubscribe to Async

### PR DESCRIPTION
**Pull request by colin-higgins**

### CODE REVIEW

**Comment by colin-higgins** [comment]
It looks like the acceptance test public async Task Should_discard_them()` is failing intermittently within `When_receiving_messages_already_dispatched` on develop.

I discovered it failing on this branch but confirmed it also failed intermittently without my changes.

@danielmarbach  @SeanFeldman

**Comment by SeanFeldman** [comment]
Looking at the threads, message pump was going on even though test should have stopped it.

**Comment by danielmarbach** [comment]
I'll look

>

-- 
*#NSBcon 2015 • The "All About NServiceBus" conference is back *
*• Learn more <http://go.particular.net/l/27442/2015-08-21/5d4wzp> →*

**Comment by danielmarbach** [comment]
Looked into it with @andreasohlund we have a proposed fix

File: src/NServiceBus.Core/IBusExtensions.cs [hunk 1]
```diff
@@ -81,7 +81,7 @@ public static void Subscribe(this IBus bus, Type messageType)
             Guard.AgainstNull("bus", bus);
             Guard.AgainstNull("messageType", messageType);
 
-            bus.Subscribe(messageType, new SubscribeOptions());
+            bus.SubscribeAsync(messageType, new SubscribeOptions());
```
Comment (danielmarbach) [inline src/NServiceBus.Core/IBusExtensions.cs]: We need to fix this. The synchronous version should be obsoleted and only the async version should be exposed

File: src/NServiceBus.Core/IBusExtensions.cs [hunk 2]
```diff
@@ -94,7 +94,7 @@ public static void Subscribe<T>(this IBus bus)
         {
             Guard.AgainstNull("bus", bus);
 
-            bus.Subscribe(typeof(T), new SubscribeOptions());
+            bus.SubscribeAsync(typeof(T), new SubscribeOptions());
```
Comment (danielmarbach) [inline src/NServiceBus.Core/IBusExtensions.cs]: We need to fix this. The synchronous version should be obsoleted and only the async version should be exposed

File: src/NServiceBus.Core/IBusExtensions.cs [hunk 3]
```diff
@@ -107,7 +107,7 @@ public static void Unsubscribe(this IBus bus, Type messageType)
             Guard.AgainstNull("bus", bus);
             Guard.AgainstNull("messageType", messageType);
 
-            bus.Unsubscribe(messageType, new UnsubscribeOptions());
+            bus.UnsubscribeAsync(messageType, new UnsubscribeOptions());
```
Comment (danielmarbach) [inline src/NServiceBus.Core/IBusExtensions.cs]: We need to fix this. The synchronous version should be obsoleted and only the async version should be exposed

File: src/NServiceBus.Core/IBusExtensions.cs [hunk 4]
```diff
@@ -119,7 +119,7 @@ public static void Unsubscribe<T>(this IBus bus)
         {
             Guard.AgainstNull("bus", bus);
 
-            bus.Unsubscribe(typeof(T), new UnsubscribeOptions());
+            bus.UnsubscribeAsync(typeof(T), new UnsubscribeOptions());
```
Comment (danielmarbach) [inline src/NServiceBus.Core/IBusExtensions.cs]: We need to fix this. The synchronous version should be obsoleted and only the async version should be exposed

File: src/NServiceBus.Core/Unicast/UnicastBusInternal_IBusImpl.cs [hunk 5]
```diff
@@ -40,21 +36,25 @@ public void Send<T>(Action<T> messageConstructor, NServiceBus.SendOptions option
...
 
         /// <inheritdoc />
-        public void Subscribe(Type eventType, SubscribeOptions options)
+        public Task SubscribeAsync(Type eventType, SubscribeOptions options)
         {
             Guard.AgainstNull("eventType", eventType);
             Guard.AgainstNull("options", options);
 
-            busImpl.Subscribe(eventType,options);
+            busImpl.SubscribeAsync(eventType,options);
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Unicast/UnicastBusInternal_IBusImpl.cs]: This should be `return busImpl.SubscribeAsync(eventType, options)`

File: src/NServiceBus.Core/Unicast/UnicastBusInternal_IBusImpl.cs [hunk 6]
```diff
@@ -40,21 +36,25 @@ public void Send<T>(Action<T> messageConstructor, NServiceBus.SendOptions option
...
 
         /// <inheritdoc />
-        public void Unsubscribe(Type eventType, UnsubscribeOptions options)
+        public Task UnsubscribeAsync(Type eventType, UnsubscribeOptions options)
         {
             Guard.AgainstNull("eventType", eventType);
             Guard.AgainstNull("options", options);
 
-            busImpl.Unsubscribe(eventType, options);
+            busImpl.UnsubscribeAsync(eventType, options);
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Unicast/UnicastBusInternal_IBusImpl.cs]: This should be `return busImpl.UnsubscribeAsync(eventType, options)`

**Comment by colin-higgins** [comment]
@danielmarbach @SeanFeldman  I have converted the extension methods to Async as well as the tests and obsoleted the Sync ones.

**Comment by danielmarbach** [comment]
@SeanFeldman can you explain to @colin-higgins rebases and that we usually prefer rebase over merge on the PRs?

File: src/NServiceBus.AcceptanceTests/ApiExtension/When_extending_the_publish_api.cs [hunk 7]
```diff
@@ -30,7 +30,7 @@ await Scenario.Define<Context>()
                      )
                     .WithEndpoint<Subscriber1>(b => b.Given((bus, context) =>
                         {
-                            bus.Subscribe<MyEvent>();
+                            bus.SubscribeAsync<MyEvent>();
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/ApiExtension/When_extending_the_publish_api.cs]: `await` missing

File: src/NServiceBus.AcceptanceTests/BestPractices/When_subscribing_to_command.cs [hunk 8]
```diff
@@ -16,7 +16,7 @@ public class When_subscribing_to_command : NServiceBusAcceptanceTest
                     {
                         try
                         {
-                            bus.Subscribe<MyCommand>();
+                            bus.SubscribeAsync<MyCommand>();
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/BestPractices/When_subscribing_to_command.cs]: `await` missing

File: src/NServiceBus.AcceptanceTests/BestPractices/When_subscribing_to_command_bestpractices_disabled_on_endpoint.cs [hunk 9]
```diff
@@ -14,7 +14,7 @@ public class When_subscribing_to_command_bestpractices_disabled_on_endpoint : NS
             await Scenario.Define<Context>()
                 .WithEndpoint<Endpoint>(b => b.Given((bus, c) =>
                 {
-                    bus.Subscribe<MyCommand>();
+                    bus.SubscribeAsync<MyCommand>();
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/BestPractices/When_subscribing_to_command_bestpractices_disabled_on_endpoint.cs]: `return bus.SubscribeAsync<MyCommand>();`

File: src/NServiceBus.AcceptanceTests/BestPractices/When_unsubscribing_to_command.cs [hunk 10]
```diff
@@ -16,7 +16,7 @@ public class When_unsubscribing_to_command : NServiceBusAcceptanceTest
                     {
                         try
                         {
-                            bus.Unsubscribe<MyCommand>();
+                            bus.UnsubscribeAsync<MyCommand>();
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/BestPractices/When_unsubscribing_to_command.cs]: `await` missing

File: src/NServiceBus.AcceptanceTests/BestPractices/When_unsubscribing_to_command_bestpractices_disabled_on_endpoint.cs [hunk 11]
```diff
@@ -14,7 +14,7 @@ public class When_unsubscribing_to_command_bestpractices_disabled_on_endpoint :
             await Scenario.Define<Context>()
                 .WithEndpoint<Endpoint>(b => b.Given((bus, c) =>
                 {
-                    bus.Unsubscribe<MyCommand>();
+                    bus.UnsubscribeAsync<MyCommand>();
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/BestPractices/When_unsubscribing_to_command_bestpractices_disabled_on_endpoint.cs]: `return   bus.UnsubscribeAsync<MyCommand>();`

File: src/NServiceBus.AcceptanceTests/Routing/When_base_event_from_2_publishers.cs [hunk 12]
```diff
@@ -20,8 +20,8 @@ public class When_base_event_from_2_publishers : NServiceBusAcceptanceTest
                      )
                .WithEndpoint<Subscriber1>(b => b.Given((bus, c) =>
                {
-                   bus.Subscribe<DerivedEvent1>();
-                   bus.Subscribe<DerivedEvent2>();
+                   bus.SubscribeAsync<DerivedEvent1>();
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Routing/When_base_event_from_2_publishers.cs]: 2 x `await` missing

File: src/NServiceBus.AcceptanceTests/Routing/When_multi_subscribing_to_a_polymorphic_event.cs [hunk 13]
```diff
@@ -26,8 +26,8 @@ public class When_multi_subscribing_to_a_polymorphic_event : NServiceBusAcceptan
                 .WithEndpoint<Subscriber1>(b => b.Given((bus, c) =>
                 {
                     c.AddTrace("Subscriber1 subscribing to both events");
-                    bus.Subscribe<IMyEvent>();
-                    bus.Subscribe<MyEvent2>();
+                    bus.SubscribeAsync<IMyEvent>();
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Routing/When_multi_subscribing_to_a_polymorphic_event.cs]: 2 x `await` missing

File: src/NServiceBus.AcceptanceTests/Routing/When_publishing.cs [hunk 14]
```diff
@@ -19,7 +19,7 @@ await Scenario.Define<Context>()
                     )
                 .WithEndpoint<Subscriber3>(b => b.Given((bus, context) =>
                 {
-                    bus.Subscribe<IFoo>();
+                    bus.SubscribeAsync<IFoo>();
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing.cs]: `await` missing

File: src/NServiceBus.AcceptanceTests/Routing/When_publishing.cs [hunk 15]
```diff
@@ -51,7 +51,7 @@ await Scenario.Define<Context>()
                      )
                     .WithEndpoint<Subscriber1>(b => b.Given((bus, context) =>
                         {
-                            bus.Subscribe<MyEvent>();
+                            bus.SubscribeAsync<MyEvent>();
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing.cs]: Please recheck all tests.

File: src/NServiceBus.Core/IBusExtensions.cs (hunk 1 above)
Comment (danielmarbach) [inline src/NServiceBus.Core/IBusExtensions.cs]: `return bus.SubscribeAsync(messageType, new SubscribeOptions());`

File: src/NServiceBus.Core/IBusExtensions.cs (hunk 2 above)
Comment (danielmarbach) [inline src/NServiceBus.Core/IBusExtensions.cs]: See above

File: src/NServiceBus.Core/IBusExtensions.cs (hunk 3 above)
Comment (danielmarbach) [inline src/NServiceBus.Core/IBusExtensions.cs]: `return bus.UnsubscribeAsync(messageType, new SubscribeOptions());`

File: src/NServiceBus.Core/IBusExtensions.cs (hunk 4 above)
Comment (danielmarbach) [inline src/NServiceBus.Core/IBusExtensions.cs]: see above

File: src/NServiceBus.Core/Routing/AutomaticSubscriptions/AutoSubscribe.cs [hunk 16]
```diff
@@ -91,7 +91,7 @@ protected override void OnStart()
             {
                 foreach (var eventType in eventsToSubscribe)
                 {
-                    bus.Subscribe(eventType);
+                    bus.SubscribeAsync(eventType);
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Routing/AutomaticSubscriptions/AutoSubscribe.cs]: `await`

File: src/NServiceBus.Core/Unicast/UnicastBusInternal_IBusImpl.cs (hunk 6 above)
Comment (danielmarbach) [inline src/NServiceBus.Core/Unicast/UnicastBusInternal_IBusImpl.cs]: Comment above still holds

File: src/NServiceBus.Core/Unicast/UnicastBusInternal_IBusImpl.cs [hunk 17]
```diff
@@ -40,21 +36,25 @@ public void Send<T>(Action<T> messageConstructor, NServiceBus.SendOptions option
...
 
         /// <inheritdoc />
-        public void Subscribe(Type eventType, SubscribeOptions options)
+        public Task SubscribeAsync(Type eventType, SubscribeOptions options)
         {
             Guard.AgainstNull("eventType", eventType);
             Guard.AgainstNull("options", options);
 
-            busImpl.Subscribe(eventType,options);
+            busImpl.SubscribeAsync(eventType, options);
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Unicast/UnicastBusInternal_IBusImpl.cs]: This should be `return bbusImpl.SubscribeAsync(eventType, options);`

**Comment by danielmarbach** [comment]
@colin-higgins why did you close this PR?

**Comment by danielmarbach** [comment]
@colin-higgins doing a rebase for you and fix a few things

**Comment by danielmarbach** [comment]
I'm pulling this in when green

**Pull request merged**